ruby_programming/automated_testing/test_driven_development.md

```diff
@@ -0,0 +1,209 @@
+### Introduction
+
+So far you’ve been learning how to test code that you’ve written, but there’s a popular development process that flips that concept on its head. Rather than writing code, manually testing it until you get it working, then writing a test to make sure it stays working; you can write the test **before** the code, so that you don’t have to waste *any* time manually testing. Test Driven Development is the name of this inverted development process.
+
+### Learning Outcomes
+Look through these now and use them to guide your learning. By the end of this lesson, expect to:
+
+- Learn what Test Driven Development is
+- Learn the advantages of Test Driven Development
+- Learn how to write code TDD style
+
+### What is TDD, and why do it?
+
+Test Driven Development (TDD) just means, in its simplest definition, that you write the tests for a method *before* the method itself. This can seem like a *really* strange concept, especially when you first try doing it, but like anything else in programming (and life more broadly) the more you do it the more normal it will feel, and the better you will get at it.
+
+Following Test Driven Development is especially useful when you’re still learning about testing, as it will help you learn how to write code that’s *easy* to test from the outset, rather than having to rewrite methods to make them testable.
+
+Another benefit of writing code TDD style is that it *guarantees* that your code will be tested. When writing code with the intention of testing it after the fact, it’s very easy to put testing on the back burner for so long that it never happens, and testing *before* you write the code in question removes that temptation. As was mentioned in the introduction, TDD’ing a project means that you don’t have to manually test much if any functionality, and that time that you would’ve spent manually testing can add up quickly on a large project. One last benefit to the TDD approach is that it can help you better plan the design, and catch bugs earlier in the process because your tests are *driving* your development.
+
+We should mention here though that not *everyone* agrees that TDD is the way to go. There are plenty of developers that write tests after the code is written. While you are free to decide how you want to write your tests in the applications you build, we strongly encourage you to follow TDD for now.
```

Comment:
I've been reading some stuff lately by  [Jason Swett](https://www.codewithjason.com/test-driven-development/)  that recommends beginners shouldn't worry too much about following TDD straight away.

I think thats something we should follow here as well. I do think it's very beneficial for students to know about TDD and its workflow but I think it could maybe be detrimental to encourage students to adopt it straight away. 

Reason being, following a TDD workflow from the start can be a barrier while students are getting comfortable with testing in general. They will also be quite early in the course with these new lessons so they will be contending with getting comfortable with coding as well.

Whats your thoughts?

Reply:
I know you are not asking my thoughts, but I agree with Jason's thoughts because that was my experience. I think it is really hard to do TDD when you are still getting to know the tool (RSpec). Since we are moving the testing section up, what do you think about keeping the current location of the Connect Four project to have the students try a project TDD?

Reply:
I'm leaning towards not enforcing tdd on any of the projects and leaving it up to the student. But if we do want to have tdd on a project, I think a smaller/simpler project would work best.

This does raise a really good question about where connect 4 should go. Lets discuss it in the discord room for the Ruby-testing project 😄

Reply:
For me, I think doing a project TDD from start to finish was super helpful in learning to test, and I would definitely like to see at least one TDD project.

I think looking at your editor and having to translate your current problem into tests, rather than only having to figure out how to test already written code, helps solidify things in your mind. I recognize that I could be alone in this opinion though. 
I also recognize that separating learning the fundamentals of testing and TDD would likely be beneficial.

If you all would like me to revise this to leave it up to the students I'm alright with that. If we think we want to keep (for example) connect four as a TDD project I would be for that as well. I think we could introduce TDD here and say students can follow it if they want, then on connect four link back to this lesson and say students should do it TDD.

Just let me know where you all want to go with it.

Reply:
For now I've revised this part of the lesson and removed the encouragement to start doing things TDD immediately. If this isn't quite what you had in mind please let me know, I will leave this unresolved for now just in case.

Reply:
Sorry for the delay in getting back to your about this @wise-king-sullyman. I agree with you and @rlmoser99 that connect 4 should be used as the project for practicing tdd. That project will come right after this lesson so it should be a nice transition.